src/player.py

```diff
@@ -713,7 +713,151 @@ def pickMove(self, game):
+        features['Has largest army'] = 1 if self.hasLargestArmy else 0
+        features['Num cards discarded'] = self.numCardsDiscarded
+        features['Score'] = self.score
+        features['Has Won'] = self.score == 10
+        features['Ratio roads to settlements'] = (len(self.roads) / numSettlements) if numSettlements > 0 else 1
+        features['Ratio cities to settlements'] = (numCities/ numSettlements) if numSettlements > 0 else 1 if numCities > 0 else 0
+        features['Squared distance to end'] = (10 - self.score)**2
+        features['Num accesible resources'] = self.getAccessibleResources(expectedResources)
+
+        #Features for other players number of each piece and total score
+        #TODO: Feature for longest road or expected resources per turn?
+        if game:
+            for player in game.players:
+                if player.turn_num != self.turn_num:
+                    features["Score " + str(player.turn_num)] = player.score
+                    features["Player "+ str(player.turn_num) + " DevCards"] = len(player.devCards) + len(player.devCardsPlayed)
+                    features["Player "+ str(player.turn_num) + " Roads"] = len(player.roads)
+                    features["Player "+ str(player.turn_num) + " Settlements"] = sum([int(i.pieceType=='Settlement') for i in player.cities_and_settlements])
+                    features["Player "+ str(player.turn_num) + " Cities"] = sum([int(i.pieceType=='City') for i in player.cities_and_settlements])
+
+        return features    
+
+    #Helper function to evaluate a gamestate
+    def eval(self,game):
+        features = self.feature_extractor(game)
+        return util.dotProduct(features, self.weights)
+
+    #Update the weights of your features after a given turn. Different from update_weights used in other AI's
+    def updateWeights(self, game):
+        # print game
+        #Get current score
+        cur_features = self.feature_extractor(game)
+        target = util.dotProduct(cur_features, self.weights)
+        pred = self.prevScore
+        
+        # print "fuck" , cur_features["Player "+ str(1) + " Settlements"]
+
+        #If there are no previous features you can't update
+        if self.prevFeatures: 
+
+            #Update weights
+            diff = pred - target
+            for feature, val in self.prevFeatures.items():
+                # print diff, val
+                self.weights[feature] -= self.eta * diff * val
+        
+        self.prevFeatures = cur_features
+        self.prevScore = target
+
+
+    #Same as superclass, but it updates weights before picking the move and stores prediction at the end of the turn
+    def pickMove(self, game):
+
+        #Update weights
+        self.updateWeights(game)
+
+        #Pick a move as normal
+        possible_moves = game.getPossibleActions(self)
+        bestMoveScore, bestMove = -10000, None
+        for possibleMove in possible_moves:
```

Comment:
Looks good! Have you tried running it against our old AI / random AI? Does it do better?

Reply:
I haven't compared them yet because I modified test and the framework is a bit different